quinn-proto/src/packet.rs

```diff
@@ -309,41 +308,53 @@ impl Packet {
 /// Protocol-level identifier for a connection.
 ///
 /// Mainly useful for identifying this connection's packets on the wire with tools like Wireshark.
-#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
-pub struct ConnectionId(pub(crate) ArrayVec<[u8; MAX_CID_SIZE]>);
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct ConnectionId {
+    pub len: u8,
+    pub bytes: [u8; MAX_CID_SIZE]
+}
 
 impl ::std::ops::Deref for ConnectionId {
     type Target = [u8];
     fn deref(&self) -> &[u8] {
-        &self.0
+        &self.bytes[0..self.len as usize]
     }
 }
 
 impl ::std::ops::DerefMut for ConnectionId {
     fn deref_mut(&mut self) -> &mut [u8] {
-        &mut self.0
+        &mut self.bytes[0..self.len as usize]
     }
 }
 
 impl ConnectionId {
-    pub fn new(data: [u8; MAX_CID_SIZE], len: usize) -> Self {
-        let mut x = ConnectionId(data.into());
-        x.0.truncate(len);
-        x
+    pub fn new(bytes: &[u8]) -> Self {
+        debug_assert!(bytes.is_empty() || (bytes.len() > 3 && bytes.len() <= MAX_CID_SIZE));
+        let mut res = Self {
+            len: bytes.len() as u8,
+            bytes: [0; MAX_CID_SIZE],
+        };
+        (&mut res.bytes[..bytes.len()]).clone_from_slice(&bytes);
+        res
     }
 
     pub fn random<R: Rng>(rng: &mut R, len: u8) -> Self {
         debug_assert!(len as usize <= MAX_CID_SIZE);
-        let mut v = ArrayVec::from([0; MAX_CID_SIZE]);
-        rng.fill_bytes(&mut v[0..len as usize]);
-        v.truncate(len as usize);
-        ConnectionId(v)
+        let mut res = Self {
+            len: len as u8,
+            bytes: [0; MAX_CID_SIZE]
+        };
+
+        let mut rng_bytes = [0; MAX_CID_SIZE];
+        rng.fill_bytes(&mut rng_bytes);
+        (&mut res.bytes[..len as usize]).clone_from_slice(&rng_bytes[..len as usize]);
+        res
     }
 }
 
 impl fmt::Display for ConnectionId {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        for byte in &self.0 {
+        for byte in &self.bytes {
```

Comment:
This, too, should be bounded by `len`, perhaps by dispatching to the `Deref` impl if that's convenient.